### PR TITLE
Simple SDK

### PR DESCRIPTION
**Pull request by sjudson**
This PR implements a basic, opinionated SDK, following https://github.com/nexus-xyz/nexus-zkvm/issues/148.

At present the PR implements a minimalistic approach, and has the following outstanding work to be taken off draft:

- [x] add support for dynamic configuration and compilation of programs
- [x] make a decision about how to handle outputs
- [x] add examples
- [x] add docs
- [x] add quickstart
- [x] add cargo setup
- [x] add support for HyperNova 
- [x] add support for Jolt

### CODE REVIEW

File: sdk/examples/nova.rs
Comment (slumber) [inline sdk/examples/nova.rs:6]: You can make it work from any dir as follows

https://github.com/nexus-xyz/nexus-zkvm/blob/3ab4d656b3f65ebb974888172438d38de8b83424/prover/examples/prove.rs#L22-L35

File: sdk/src/traits.rs
Comment (slumber) [inline sdk/src/traits.rs:19]: Should it be autoimplemented with `fs::read()` + `Self::new()`?

File: sdk/examples/nova.rs
Comment (slumber) [inline sdk/examples/nova.rs:16]: Missing input?

**Review by slumber — COMMENTED** [review]
Like how neat the example of using nova is, especially compared to `nexus-prover`

File: sdk/examples/nova.rs
Comment (sjudson) [reply sdk/examples/nova.rs:16]: Ah, yeah, forgot to add (program still runs of course).

File: sdk/src/compile.rs
Comment (slumber) [inline sdk/src/compile.rs:166]: While scripts are similar, I still think `include_str` is easier to maintain

**Review by slumber — COMMENTED** [review]
The review is quite pedantic since it's a user-facing crate we plan to publish.


Overall, nicely done.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:38]: I didn't want users to use `release` and then have it blowup, was the reasoning between limiting it to `release-unoptimized` or `debug`, at least until https://github.com/nexus-xyz/nexus-zkvm/issues/155 is resolved.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:39]: Fair point. I may leave this code in but commented out with a note that this is waiting on `native` build feature parity.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:35]: I think in general there are a lot of enhancements that could be made here, but wanted something simple and clean to get it out the door.

I can open an enhancement issue mentioning some of this stuff once this merges.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:129]: As noted, will comment out.

File: sdk/src/hypernova/seq.rs
Comment (sjudson) [reply sdk/src/hypernova/seq.rs:88]: Oh sorry, this is a leftover from when I was trying to shoehorn `jolt` into the `Prover` trait. Will replace.

File: sdk/src/traits.rs
Comment (sjudson) [reply sdk/src/traits.rs:53]: The main `compile` functionality is detached from the prover, but compilation has enough prover dependent elements (linker script, memory limit) that the compilation needs to know for what prover it is building towards.

Pushing it through the object seemed more misuse resistant than relying on the user configuring it correctly.

File: sdk/src/traits.rs
Comment (sjudson) [reply sdk/src/traits.rs:59]: I generally prefer explicitness with this type of thing. I'd prefer to start with this interface, and we can add others if users ask for them.

File: sdk/src/compile.rs
Comment (slumber) [reply sdk/src/compile.rs:35]: As long as it's not published as `1.0.0`. Otherwise, it will be a breaking change.

Comment (sjudson) [reply sdk/src/compile.rs:35]: I mean I think it's fine for us to have a major bump of the `sdk` at some point in the future as we steer towards a more production-ready deployment.

File: sdk/src/traits.rs
Comment (slumber) [reply sdk/src/traits.rs:59]: I imagine a lot of people will stumble upon
```rust
<P as Prover>::run(prover, None).unwrap(); // compile error, must specify T
```

Comment (sjudson) [reply sdk/src/traits.rs:59]: I mean I can push RTFM hard: make sure it's in the quickstart and in the template.

File: sdk/src/hypernova/seq.rs
Comment (sjudson) [reply sdk/src/hypernova/seq.rs]: Wait, I misremembered. This is actually to allow the return type of the `prove` function to be parameterized by `U` without requiring the entire trait to be parameterized by it (which I thought was unnecessarily confusing).

Comment (slumber) [reply sdk/src/hypernova/seq.rs]: It will work if we agree on https://github.com/nexus-xyz/nexus-zkvm/pull/208#discussion_r1673875314

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:35]: See also: https://github.com/nexus-xyz/nexus-zkvm/pull/208#discussion_r1674445747

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:37]: Decided to make `new()` just take the package name, and added `new_with_custom_binary()` that takes package + binary, to make the setup even cleaner in the most likely case.

File: sdk/src/traits.rs
Comment (sjudson) [reply sdk/src/traits.rs:17]: It is concatenated, though I could change that.

This is somewhat of an interesting philosophical question. Should the verifier be able to see which `ecall` produced which amendment to the logs? 

I'm not sure, but given that provides strictly less information, I'll convert this to provide a `Vec`.

File: sdk/src/traits.rs
Comment (sjudson) [reply sdk/src/traits.rs:15]: I debated this interface initially. In general, I think I pretty strongly prefer the single specification of the input/output interface at run. However, as you note, there are plenty of knock-on effects for having that. 

I'll add one more though, that I think just caused me to agree with you that this sort of approach is the better way to go. It's possible that the host won't necessarily know the appropriately `U`, since different computational paths of the guest program might write different output types to the tape. If using the single interface, the host would have to rerun the entire computation to "test" a different `U`, whereas using this interface they can invoke `output` multiple times with different `U` until one doesn't fail.

Comment (slumber) [reply sdk/src/traits.rs:15]: This is what my comment says: "A simple use case is serializing different types based on the input. "

Comment (sjudson) [reply sdk/src/traits.rs:15]: Ah sorry, didn't process that.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:111]: I'm pretty sure it sets the size of the memory usable by the program, as well as the bytecode (since it's the same memory view). The assertion here is just scoped to capturing too large of a program in the original setup.

File: sdk/src/compile.rs
Comment (sjudson) [reply sdk/src/compile.rs:92]: Macro will not be overwritten by `sdk`, which I can document. And I'm fine with  `memlimit` as the user facing name, and `stack_size` as the somewhat more descriptive internal name.

I can investigate what might be causing this to behave strangely at present before merging...

File: tools/src/command/host.rs
Comment (yoichi-nexus) [inline tools/src/command/host.rs:49]: I had to add
```
[workspace]
members = [
    "src/guest"
]
```
manually before getting `cargo run -r` working.

Comment (sjudson) [reply tools/src/command/host.rs:49]: Oh interesting. Maybe cargo version? Will add.

File: runtime/src/ecalls.rs
Comment (slumber) [inline runtime/src/ecalls.rs:36]: nits:
1. Document that it exhausts the private tape, and essentially can only be called once.
2.
```rust
let bytes: Vec<u8> = core::iter::from_fn(read_from_private_input).collect();
postcard::from_bytes::<T>(bytes.as_slice())
```
to have a single implementation wrapping the ecall
3. Let's add native counterparts so that compilation at least doesn't fail.

**Review by yoichi-nexus — COMMENTED** [review]
When I follow the instructions on the branch now I get
```
warning: compiling this new package may not work due to invalid workspace configuration

failed to parse manifest at `/Users/yh/shared/nexus/nexus-zkvm/nexus-host/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  `workspace.package.edition` was not defined
note: see more `Cargo.toml` keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
error: failed to parse manifest at `/Users/yh/shared/nexus/nexus-zkvm/nexus-host/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  `workspace.package.edition` was not defined
Error: cargo didn't exit successfully: exit status: 101
```

I don't know if this is an issue of my environment of an issue of the branch.

**Comment by sjudson** [comment]
@yoichi-nexus I can't reproduce on my end.

**Comment by yoichi-nexus** [comment]
> @yoichi-nexus I can't reproduce on my end.

Then this doesn't block this PR, and I have to figure this out on my end.

**Review by nexodon — APPROVED** [review]
lgtm

File: runtime/src/ecalls.rs
Comment (slumber) [inline runtime/src/ecalls.rs:39]: Typed `read` exhausts it, but this function doesn't

File: sdk/Cargo.toml
Comment (slumber) [inline sdk/Cargo.toml:21]: nit newline

File: sdk/README.md
Comment (slumber) [inline sdk/README.md:189]: typo

File: sdk/src/compile.rs
Comment (slumber) [inline sdk/src/compile.rs:49]: a note that it's not a big change to detect binary name from cargo manifest, as the error returned to user is a unit enum without a message.

Can be implemented in 0.2.1

File: tools/src/command/host.rs
Comment (slumber) [inline tools/src/command/host.rs:38]: needs an update before merge

File: tools/src/command/host.rs
Comment (slumber) [inline tools/src/command/host.rs:106]: I think it's better to re-export from sdk or wrap it into public functions.

Overall I think the host functionality is not going to be easy to maintain, but here's the best effort.

**Review by slumber — COMMENTED** [review]
Minor comments, mainly no blockers. Once resolved, can be merged since there's an approve already

**Pull request merged**
